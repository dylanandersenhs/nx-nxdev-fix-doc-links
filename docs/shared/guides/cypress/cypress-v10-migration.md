# Migrating to Cypress V10

# TODO(caleb): update this when cypress migration is available

Cypress v10 introduce new features, like component testing, along with some breaking changes.

Before continuing, make sure you have all your changes committed and have a clean working tree.

You can migrate an E2E project to v10 by running the following command:

```bash
nx g @nrwl/cypress:convert-cypress-ten --project=<e2e-project-to-convert>
```

Optionally, you can convert all e2e projects with the `--all` flag:

```bash
nx g @nrwl/cypress:convert-cypress-ten --all
```

We take the best effort to make this migration seamless, but there can be edge cases we didn't anticipate.
So consulting the [Cypress migration guide](https://docs.cypress.io/guides/guides/cypress-migration-guide.html) is
highly recommended.

In general, these are the steps taken to migrate your project:

1. create a new `cypress.config.ts` file based on the existing `cypress.json` file
   - the `pluginsFile` option has been replaced for `setupNodeEvents`. We will import the file and add it to
     the `setupNodeEvents` config option. Double check your plugins are working correctly.
2. rename all test files from `spec.ts` to `cy.ts`
3. rename `support/index.ts` to `support/e2e.ts`
4. rename `integrations` folder to `e2e` folder

The Cypress migration guide [contains more details](https://docs.cypress.io/guides/guides/cypress-migration-guide.html)
and will contain the most update to date information.
