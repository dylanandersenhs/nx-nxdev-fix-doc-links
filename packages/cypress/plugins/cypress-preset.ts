import { workspaceRoot } from '@nrwl/devkit';
import { join, relative } from 'path';

interface BaseCypressPreset {
  videosFolder: string;
  screenshotsFolder: string;
  video: boolean;
  chromeWebSecurity: boolean;
}

export function nxBaseCypressPreset(pathToConfig: string): BaseCypressPreset {
  const projectPath = relative(workspaceRoot, pathToConfig);
  const offset = relative(pathToConfig, workspaceRoot);
  const videosFolder = join(offset, 'dist', 'cypress', projectPath, 'videos');
  const screenshotsFolder = join(
    offset,
    'dist',
    'cypress',
    projectPath,
    'screenshots'
  );

  return {
    videosFolder,
    screenshotsFolder,
    video: true,
    chromeWebSecurity: false,
  };
}

/**
 * nx E2E Preset for Cypress
 * @description
 * this preset contains the base configuration
 * for your e2e tests that nx recommends.
 * you can easily extend this within your cypress config via spreading the preset
 * @example
 * export default defineConfig({
 *   e2e: {
 *     ...nxE2EPreset(__dirname)
 *     // add your own config here
 *   }
 * })
 *
 * @param pathToConfig will be used to construct the output paths for videos and screenshots
 */
export function nxE2EPreset(pathToConfig: string) {
  return {
    ...nxBaseCypressPreset(pathToConfig),
    fileServerFolder: '.',
    supportFile: 'src/support/e2e.ts',
    specPattern: 'src/**/*.cy.{js,jsx,ts,tsx}',
    fixturesFolder: 'src/fixtures',
  };
}
