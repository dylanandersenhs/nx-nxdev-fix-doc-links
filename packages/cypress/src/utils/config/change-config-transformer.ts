import {
  CallExpression,
  createPrinter,
  createSourceFile,
  Expression,
  Identifier,
  isCallExpression,
  isExportAssignment,
  isNumericLiteral,
  isObjectLiteralExpression,
  isPropertyAssignment,
  isSpreadAssignment,
  Node,
  NodeFactory,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
  ScriptTarget,
  SourceFile,
  StringLiteral,
  SyntaxKind,
  transform,
  TransformationContext,
  TransformerFactory,
  visitEachChild,
  visitNode,
  Visitor,
} from 'typescript';
import { InternalResolvedConfigOptions } from './cypress-config.model';
import {
  CypressConfigPropertyPath,
  isBooleanLiteral,
  Overwrite,
} from './transformer.helper';

export type ModifiedCypressConfig = Partial<
  Overwrite<
    InternalResolvedConfigOptions,
    {
      component?: Overwrite<
        InternalResolvedConfigOptions['component'],
        { devServer?: { tsConfig: string; compiler: string } }
      >;
    }
  >
>;

type PrimitiveValue = string | number | boolean;
type DevServer = {
  [key in 'tsConfig' | 'compiler']?: PrimitiveValue;
};
type UpsertArgs = {
  type: 'upsert';
  newConfig: ModifiedCypressConfig;
  overwrite?: boolean;
};
type DeleteArgs = { type: 'delete' };

const devServerPositionalArgument = ['tsConfig', 'compiler'] as const;

/**
 * Update cypress.config.ts file properties.
 * Does not support cypress.config.json. use updateJson from @nrwl/devkit
 */
export class CypressConfigTransformer {
  private static configMetadataMap = new Map<
    string,
    PrimitiveValue | Map<string, PrimitiveValue | DevServer>
  >();

  private static propertiesToRemove: CypressConfigPropertyPath[] = [];

  private static sourceFile: SourceFile;

  static removeProperties(
    existingConfigContent: string,
    toRemove: CypressConfigPropertyPath[]
  ): string {
    this.configMetadataMap = new Map();
    this.propertiesToRemove = toRemove;

    this.sourceFile = createSourceFile(
      'cypress.config.ts',
      existingConfigContent,
      ScriptTarget.Latest,
      true
    );

    const transformedResult = transform(this.sourceFile, [
      this.changePropertiesTransformer({ type: 'delete' }),
    ]);

    return createPrinter().printFile(transformedResult.transformed[0]);
  }

  static addOrUpdateProperties(
    existingConfigContent: string,
    newConfig: ModifiedCypressConfig,
    overwrite = false
  ): string {
    this.configMetadataMap = new Map();
    this.propertiesToRemove = [];

    this.sourceFile = createSourceFile(
      'cypress.config.ts',
      existingConfigContent,
      ScriptTarget.Latest,
      true
    );

    const transformedResult = transform(this.sourceFile, [
      this.changePropertiesTransformer({
        type: 'upsert',
        newConfig,
        overwrite,
      }),
    ]);

    return createPrinter().printFile(transformedResult.transformed[0]);
  }

  private static changePropertiesTransformer(
    change: UpsertArgs | DeleteArgs
  ): TransformerFactory<SourceFile> {
    return (context: TransformationContext) => {
      if (change.type === 'upsert') {
        // before visiting the sourceFile (aka the existing config)
        // we add the newConfig, as TypeScript AST, to our ConfigMetadata
        const newConfigAst: ObjectLiteralExpression =
          context.factory.createObjectLiteralExpression(
            Object.entries(change.newConfig).map(([configKey, configValue]) =>
              createObjectAssignments(context.factory, configKey, configValue)
            ),
            true
          );

        this.buildMetadataFromConfig(context.factory, newConfigAst);
      }
      return (sourceFile: SourceFile) => {
        const nodeVisitor: Visitor = (node: Node): Node => {
          if (isExportAssignment(node)) {
            const callExpression = node.expression as CallExpression;
            const rootConfigNode = callExpression
              .arguments[0] as ObjectLiteralExpression;

            if (
              change.type === 'delete' ||
              (change.type === 'upsert' && !change.overwrite)
            ) {
              this.buildMetadataFromConfig(context.factory, rootConfigNode);
            }

            return context.factory.updateExportAssignment(
              node,
              node.decorators,
              node.modifiers,
              context.factory.updateCallExpression(
                callExpression,
                callExpression.expression,
                callExpression.typeArguments,
                [
                  context.factory.createObjectLiteralExpression(
                    [...this.configMetadataMap.entries()]
                      .map(([configKey, configValue]) => {
                        return createObjectAssignments(
                          context.factory,
                          configKey,
                          configValue
                        );
                      })
                      .sort((propA, propB) =>
                        isSpreadAssignment(propA) ? -1 : 1
                      ),
                    true
                  ),
                ]
              )
            );
          }

          return visitEachChild(node, nodeVisitor, context);
        };

        return visitNode(sourceFile, nodeVisitor);
      };
    };
  }

  private static buildMetadataFromConfig(
    factory: NodeFactory,
    config: ObjectLiteralExpression,
    metadataMap = this.configMetadataMap,
    propertiesToRemove: CypressConfigPropertyPath[] = this.propertiesToRemove,
    parentPrefix = ''
  ): void {
    if (!Array.isArray(config.properties)) {
      console.log(config);
      return;
    }

    for (const property of config.properties) {
      let assignment = property;
      if (isPropertyAssignment(property)) {
        const assignmentName = (assignment.name as Identifier).text;
        const propertyPath = parentPrefix
          ? parentPrefix.concat('.', assignmentName)
          : assignmentName;

        if (
          propertiesToRemove.includes(propertyPath as CypressConfigPropertyPath)
        ) {
          continue;
        }

        if (assignmentName === 'devServer') {
          if (isCallExpression(assignment.initializer)) {
            assignment = factory.updatePropertyAssignment(
              assignment,
              assignment.name,
              factory.createObjectLiteralExpression(
                assignment.initializer.arguments.map((arg, index) => {
                  return factory.createPropertyAssignment(
                    factory.createIdentifier(
                      devServerPositionalArgument[index]
                    ),
                    arg
                  );
                }),
                true
              )
            );
          }
        }

        const existingMetadata = metadataMap.get(assignmentName);
        if (existingMetadata !== undefined) {
          if (existingMetadata instanceof Map) {
            if (isCallExpression(assignment.initializer)) {
              existingMetadata.set(
                assignment.initializer.expression.getFullText(),
                [
                  // we are in an existing object so now we need to spread the existing object
                  // i.e. { blah: somFn(args)} => { blah: {...someFn(args), anotherProp} }
                  SyntaxKind.SpreadAssignment,
                  assignment.initializer.arguments,
                ] as any
              );
            } else {
              this.buildMetadataFromConfig(
                factory,
                assignment.initializer as ObjectLiteralExpression,
                existingMetadata as any,
                propertiesToRemove,
                propertyPath
              );
            }
          }
          continue;
        }

        if (isObjectLiteralExpression(assignment.initializer)) {
          const childMetadataMap = new Map();
          metadataMap.set(assignmentName, childMetadataMap);
          this.buildMetadataFromConfig(
            factory,
            assignment.initializer,
            childMetadataMap,
            propertiesToRemove,
            propertyPath
          );
        } else if (isCallExpression(assignment.initializer)) {
          metadataMap.set(assignmentName, [
            assignment.initializer.kind,
            assignment.initializer.expression.getFullText(),
            assignment.initializer.arguments,
          ] as any);
        } else {
          metadataMap.set(
            assignmentName,
            fromLiteralToPrimitive(assignment.initializer)
          );
        }
      } else if (isSpreadAssignment(property)) {
        if (isCallExpression(property.expression)) {
          const callExpression = property.expression;
          metadataMap.set(callExpression.expression.getFullText(), [
            SyntaxKind.SpreadAssignment,
            callExpression.arguments,
          ] as any);
        }
      }
    }
  }
}

function fromLiteralToPrimitive(nodeInitializer: Expression): PrimitiveValue {
  if (isNumericLiteral(nodeInitializer)) {
    return Number(nodeInitializer.text);
  }

  if (isBooleanLiteral(nodeInitializer)) {
    if (nodeInitializer.kind === SyntaxKind.TrueKeyword) {
      return true;
    }

    if (nodeInitializer.kind === SyntaxKind.FalseKeyword) {
      return false;
    }
  }

  return (nodeInitializer as StringLiteral).text;
}

function createObjectAssignments(
  factory: NodeFactory,
  key: string,
  value:
    | unknown
    | [type: SyntaxKind, args: Expression[]]
    | [type: SyntaxKind, identifier: string, args: Expression[]]
): ObjectLiteralElementLike {
  if (key === 'devServer' && value instanceof Map) {
    return factory.createPropertyAssignment(
      factory.createIdentifier('devServer'),
      factory.createCallExpression(
        factory.createIdentifier('componentDevServer'),
        undefined,
        // TODO(caleb): parse args into correct types
        //  if we use anything other than string down the road
        Array.from(value.values()).map((v) =>
          factory.createStringLiteral(v, true)
        )
      )
    );
  }

  if (Array.isArray(value)) {
    if (value[0] === SyntaxKind.CallExpression) {
      // this handle the case where the property assignment is a fn call;
      return factory.createPropertyAssignment(
        factory.createIdentifier(key),
        factory.createCallExpression(
          factory.createIdentifier(value[1]),
          undefined,
          value[2]
        )
      );
    } else if (value[0] === SyntaxKind.SpreadAssignment) {
      return factory.createSpreadAssignment(
        factory.createCallExpression(
          factory.createIdentifier(key),
          undefined,
          value[1]
        )
      );
    }
  }

  switch (typeof value) {
    case 'number':
      return factory.createPropertyAssignment(
        factory.createIdentifier(key),
        factory.createNumericLiteral(value)
      );
    case 'string':
      return factory.createPropertyAssignment(
        factory.createIdentifier(key),
        factory.createStringLiteral(value, true)
      );
    case 'boolean':
      return factory.createPropertyAssignment(
        factory.createIdentifier(key),
        value ? factory.createTrue() : factory.createFalse()
      );

    case 'object':
      let configEntries = Object.entries(value);

      if (value instanceof Map) {
        configEntries = Array.from(value.entries());
      }

      return factory.createPropertyAssignment(
        factory.createIdentifier(key),
        factory.createObjectLiteralExpression(
          configEntries
            .map(([configKey, configValue]) => {
              return createObjectAssignments(factory, configKey, configValue);
            })
            // make sure spread assignments go first, so they don't override the properties.
            .sort((propA, propB) => (isSpreadAssignment(propA) ? -1 : 1)),
          true
        )
      );
  }
}
