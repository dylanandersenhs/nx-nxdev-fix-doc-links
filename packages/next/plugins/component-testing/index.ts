import { nxBaseCypressPreset } from '@nrwl/cypress/plugins/cypress-preset';
import { workspaceLayout } from 'nx/src/project-graph/file-utils';
import { workspaceRoot } from 'nx/src/utils/app-root';
import { join, resolve } from 'path';
// doesn't like ESM imports
const findNextWebpackConfig = require('@cypress/react/plugins/next/findNextWebpackConfig');

/**
 * nx Component Testing Preset for Cypress
 * @description
 * this preset contains the base configuration
 * for your component tests that nx recommends.
 * including a devServer that supports nx workspaces.
 * you can easily extend this within your cypress config via spreading the preset
 * @example
 * export default defineConfig({
 *   component: {
 *     ...nxComponentTestingPreset(__dirname)
 *     // add your own config here
 *   }
 * })
 *
 * @param pathToConfig will be used to construct the output paths for videos and screenshots
 */
export function nxComponentTestingPreset(pathToConfig: string) {
  return {
    ...nxBaseCypressPreset(pathToConfig),
    devServer: componentDevServer(),
  };
}

/**
 * Start the devserver for Cypress component tests with a NextJS project.
 */
export function componentDevServer() {
  return async (cypressConfigOptions) => {
    // when getting the next config it will look for a next.config.js file and run it.
    // this will run the withNx() plugin which adds support for css modules and libs.
    // also allows for extending the webpack config as needed.
    const webpackConfig = await findNextWebpackConfig(
      cypressConfigOptions.config
    );

    if (!webpackConfig) {
      throw new Error(
        'Could not find a webpack config for next. is @nrwl/next installed?'
      );
    }
    const nextLoaders = webpackConfig.module.rules.find(
      (rule: any) => typeof rule.oneOf === 'object'
    );

    const moduleLoaders = nextLoaders.oneOf.filter((rule) =>
      regexEqual(rule.test, /\.(tsx|ts|js|cjs|mjs|jsx)$/)
    );

    if (moduleLoaders) {
      const { libsDir } = workspaceLayout();
      const libPath = join(workspaceRoot, libsDir);

      for (const moduleLoader of moduleLoaders) {
        moduleLoader.include.push(libPath);
      }
    }

    // needs to be imported after loading the webpack config from next;
    const { startDevServer } = require('@cypress/webpack-dev-server');

    return startDevServer({
      options: cypressConfigOptions,
      webpackConfig,
      template: resolve(__dirname, 'next.template.html'),
    });
  };
}

function regexEqual(x, y) {
  return (
    x instanceof RegExp &&
    y instanceof RegExp &&
    x.source === y.source &&
    x.global === y.global &&
    x.ignoreCase === y.ignoreCase &&
    x.multiline === y.multiline
  );
}
