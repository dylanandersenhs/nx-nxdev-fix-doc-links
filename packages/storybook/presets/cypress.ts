import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';

export function nxE2EStorybookPreset(filePath: string) {
  return {
    ...nxE2EPreset(filePath),
    baseUrl: 'http://localhost:4400',
  };
}
