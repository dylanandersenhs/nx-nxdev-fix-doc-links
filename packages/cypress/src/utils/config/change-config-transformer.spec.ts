import { CypressConfigTransformer } from './change-config-transformer';

describe('Update Cypress Config', () => {
  const defaultConfigContent = `
import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


export default defineConfig({
  component: nxComponentTestingPreset(__dirname),
  e2e: nxE2EPreset(__dirname),
})`;

  const configWithSpread = `
import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


export default defineConfig({
  component: {
    ...nxComponentTestingPreset(__dirname),
    video: false,
    screenshotsFolder: '../blah/another/value'
  }
  e2e: {
    ...nxE2EPreset(__dirname),
    video: false,
    screenshotsFolder: '../blah/another/value'
  }
})`;

  const expandedConfigContent = `
import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';


export default defineConfig({
  baseUrl: 'blah its me',
  component: {
    devServer: componentDevServer('tsconfig.cy.json', 'babel'),
    video: true,
    chromeWebSecurity: false,
    fixturesFolder: 'cypress/fixtures',
    specPattern: '**/*.cy.{js,jsx,ts,tsx}',
    supportFile: 'cypress/support/component.ts',
    videosFolder: '../../dist/cypress/apps/n/videos',
    screenshotsFolder: '../../dist/cypress/apps/n/screenshots',
  },
  e2e: {
    fileServerFolder: '.',
    fixturesFolder: './src/fixtures',
    integrationFolder: './src/e2e',
    supportFile: './src/support/e2e.ts',
    specPattern: '**/*.cy.{js,ts}',
    video: true,
    videosFolder: '../../dist/cypress/apps/myapp4299814-e2e/videos',
    screenshotsFolder: '../../dist/cypress/apps/myapp4299814-e2e/screenshots',
    chromeWebSecurity: false,
  }
});
  `;

  it('should add and update existing properties', () => {
    const actual = CypressConfigTransformer.addOrUpdateProperties(
      expandedConfigContent,
      {
        blah: 'i am a top level property',
        baseUrl: 'http://localhost:1234',
        component: {
          fixturesFolder: 'cypress/fixtures/cool',
          devServer: { tsConfig: 'tsconfig.spec.json', compiler: 'swc' },
          // @ts-ignore
          blah: 'i am a random property',
        },
        e2e: {
          video: false,
        },
      }
    );

    expect(actual).toMatchSnapshot();
  });

  it('should overwrite existing config', () => {
    const actual = CypressConfigTransformer.addOrUpdateProperties(
      expandedConfigContent,
      {
        baseUrl: 'http://overwrite:8080',
        component: {
          devServer: { tsConfig: 'tsconfig.spec.json', compiler: 'swc' },
        },
        e2e: {
          video: false,
        },
      },
      true
    );

    expect(actual).toMatchSnapshot();
  });

  it('should remove properties', () => {
    const actual = CypressConfigTransformer.removeProperties(
      expandedConfigContent,
      [
        'baseUrl',
        'component.devServer',
        'component.specPattern',
        'component.video',
        'e2e.chromeWebSecurity',
        'e2e.screenshotsFolder',
        'e2e.video',
      ]
    );
    expect(actual).toMatchSnapshot();
  });

  it('should add property to default config', () => {
    const actual = CypressConfigTransformer.addOrUpdateProperties(
      defaultConfigContent,
      {
        e2e: {
          baseUrl: 'http://localhost:1234',
        },
        component: {
          video: false,
        },
      }
    );

    expect(actual).toMatchSnapshot();
  });

  it('should add property with spread config', () => {
    const actual = CypressConfigTransformer.addOrUpdateProperties(
      configWithSpread,
      {
        e2e: {
          baseUrl: 'http://localhost:1234',
        },
        component: {
          defaultCommandTimeout: 60000,
        },
      }
    );

    expect(actual).toMatchSnapshot();
  });

  it('should delete a property with spread config', () => {
    const actual = CypressConfigTransformer.removeProperties(configWithSpread, [
      'component.defaultCommandTimeout',
      'component.screenshotsFolder',
      'e2e.baseUrl',
      'e2e.video',
    ]);

    expect(actual).toMatchSnapshot();
  });

  it('should not change the default config with removal', () => {
    // default config is a direct assignment vs object expression so there is nothing to remove.
    const actual = CypressConfigTransformer.removeProperties(
      defaultConfigContent,
      [
        'component.screenshotsFolder',
        'component.video',
        'e2e.screenshotsFolder',
        'e2e.video',
      ]
    );

    expect(actual).toMatchSnapshot();
  });
});
