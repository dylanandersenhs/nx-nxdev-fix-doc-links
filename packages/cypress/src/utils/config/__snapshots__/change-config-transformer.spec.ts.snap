// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Cypress Config should add and update existing properties 1`] = `
"import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';
export default defineConfig({
    blah: 'i am a top level property',
    baseUrl: 'http://localhost:1234',
    component: {
        fixturesFolder: 'cypress/fixtures/cool',
        devServer: componentDevServer('tsconfig.spec.json', 'swc'),
        blah: 'i am a random property',
        video: true,
        chromeWebSecurity: false,
        specPattern: '**/*.cy.{js,jsx,ts,tsx}',
        supportFile: 'cypress/support/component.ts',
        videosFolder: '../../dist/cypress/apps/n/videos',
        screenshotsFolder: '../../dist/cypress/apps/n/screenshots'
    },
    e2e: {
        video: false,
        fileServerFolder: '.',
        fixturesFolder: './src/fixtures',
        integrationFolder: './src/e2e',
        supportFile: './src/support/e2e.ts',
        specPattern: '**/*.cy.{js,ts}',
        videosFolder: '../../dist/cypress/apps/myapp4299814-e2e/videos',
        screenshotsFolder: '../../dist/cypress/apps/myapp4299814-e2e/screenshots',
        chromeWebSecurity: false
    }
});
"
`;

exports[`Update Cypress Config should add property to default config 1`] = `
"import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';
export default defineConfig({
    e2e: {
        ... nxE2EPreset(__dirname),
        baseUrl: 'http://localhost:1234'
    },
    component: {
        ... nxComponentTestingPreset(__dirname),
        video: false
    }
});
"
`;

exports[`Update Cypress Config should add property with spread config 1`] = `
"import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';
export default defineConfig({
    e2e: {
        ...nxE2EPreset(__dirname),
        baseUrl: 'http://localhost:1234',
        video: false,
        screenshotsFolder: '../blah/another/value'
    },
    component: {
        ...nxComponentTestingPreset(__dirname),
        defaultCommandTimeout: 60000,
        video: false,
        screenshotsFolder: '../blah/another/value'
    }
});
"
`;

exports[`Update Cypress Config should delete a property with spread config 1`] = `
"import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';
export default defineConfig({
    component: {
        ...nxComponentTestingPreset(__dirname),
        video: false
    },
    e2e: {
        ...nxE2EPreset(__dirname),
        screenshotsFolder: '../blah/another/value'
    }
});
"
`;

exports[`Update Cypress Config should not change the default config with removal 1`] = `
"import { defineConfig } from 'cypress';
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';
export default defineConfig({
    component:  nxComponentTestingPreset(__dirname),
    e2e:  nxE2EPreset(__dirname)
});
"
`;

exports[`Update Cypress Config should overwrite existing config 1`] = `
"import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';
export default defineConfig({
    baseUrl: 'http://overwrite:8080',
    component: {
        devServer: componentDevServer('tsconfig.spec.json', 'swc')
    },
    e2e: {
        video: false
    }
});
"
`;

exports[`Update Cypress Config should remove properties 1`] = `
"import { defineConfig } from 'cypress';
import { componentDevServer } from '@nrwl/cypress/plugins/next';
export default defineConfig({
    component: {
        chromeWebSecurity: false,
        fixturesFolder: 'cypress/fixtures',
        supportFile: 'cypress/support/component.ts',
        videosFolder: '../../dist/cypress/apps/n/videos',
        screenshotsFolder: '../../dist/cypress/apps/n/screenshots'
    },
    e2e: {
        fileServerFolder: '.',
        fixturesFolder: './src/fixtures',
        integrationFolder: './src/e2e',
        supportFile: './src/support/e2e.ts',
        specPattern: '**/*.cy.{js,ts}',
        videosFolder: '../../dist/cypress/apps/myapp4299814-e2e/videos'
    }
});
"
`;
